### DISCUSSION TITLE
Add an example of how to do cross-compilation without qemu on a single host

### DISCUSSION THREAD

**Discussion by voltagex** [docker/buildx#696] [Q&A]
The README states that I can `use a stage in Dockerfile to cross-compile to different architectures.`, however I can't find any examples of doing this - everything either uses SSH to another machine or qemu. I have already put work into cross-compiling various software, I'd rather just be able to do that without adding more moving parts.

Are there any examples of doing this? Can I select a different image per arch, or do I need to have a single image with multiple cross compilers, then use the BUILDARCH variable or something?

**Comment by morlay** [comment]
> do cross-compilation without qemu on a single host

Use the `FROM --platform ${BUILDPLATFORM}`, example golang

```Dockerfile
FROM --platform ${BUILDPLATFORM} golang AS builder
# now you could do cross-compilation without qemu and the architecture will following host

ARG TARGETARCH
RUN GOARCH=${TARGETARCH} go build -o /bin/app  ./cmd/app

FROM scrach 
# the stage need qemu to make builder support to differect platform from platform of host .
# if some RUN here need qemu too

COPY --from=builder /bin/app /bin/app
```


> Can I select a different image per arch

Yes. you could split to multiple stages for per arch

```Dockerfile
FROM image-for-amd64 AS builder-amd64
FROM image-for-arm64 AS builder-arm64

FROM builder-${TARGETARCH}
```

**Comment by tonistiigi** [comment]
More advanced cross-compilation cases are described in https://github.com/tonistiigi/xx#dockerfile-cross-compilation-primer  .

**Reply by voltagex** [reply]
Thanks - this is cool, but still appears to rely on qemu rather than native cross-compilers.

**Reply by tonistiigi** [reply]
No, it does not. What part is confusing you?

**Reply by baxterworks** [reply]
Apologies, I misread the link at the bottom of the README as it using binfmt, rather than compiling it. I'll check it out, thanks.
